### ISSUE TITLE
Declarative routing

### ISSUE THREAD

**Issue by jakearchibald** [w3c/ServiceWorker#1373]
Here are the requirements I'm working towards:

* Be able to bypass the service worker for particular requests.
* Speed up simple offline-first, online-first routes by avoiding service worker startup time.
* Be polyfillable – do not introduce things that cannot already be done in a fetch event.
* Be extensible – consider what future additions to the API might look like.
* Avoid state on the registration if possible – prefer state on the service worker itself.

I'm going to start with static routes, and provide additional ideas in follow-up posts.

The aim is to allow the developer to declaratively express a series of steps the browser should perform in attempt to get a response.

# Creating a route

```webidl
partial interface ServiceWorkerGlobalScope {
  attribute ServiceWorkerRouter router;
}

[Exposed=ServiceWorker]
interface ServiceWorkerRouter {
  void add(ServiceWorkerRouterItem... items);
}

[Exposed=ServiceWorker]
interface ServiceWorkerRouterItem {}
```

```js
router.add(...items);
router.add(...otherItems);
```

The browser will consider routes in the order declared, and will consider route items in the order they're given.

# Route items

Route items fall into two categories:

* Conditions – These determine if additional items should be considered.
* Sources – A place to attempt to get a response from.

## Sources

```webidl
[Exposed=ServiceWorker, Constructor(optional RouterSourceNetworkOptions options)]
interface RouterSourceNetwork : ServiceWorkerRouterItem {}

dictionary RouterSourceNetworkOptions {
  // A specific request can be provided, otherwise the current request is used.
  Request request;
}

[Exposed=ServiceWorker, Constructor(optional RouterSourceCacheOptions options)]
interface RouterSourceCache : ServiceWorkerRouterItem {}

RouterSourceCacheOptions : MultiCacheQueryOptions {
  // A specific request can be provided, otherwise the current request is used.
  Request request;
}

[Exposed=ServiceWorker, Constructor(optional RouterSourceFetchEventOptions options)]
interface RouterSourceFetchEvent : ServiceWorkerRouterItem {}

dictionary RouterSourceFetchEventOptions {
  DOMString id = '';
}
```

These interfaces don't have attributes, but they could reflect the options/defaults passed into the constructor.

## Conditions

```webidl
[Exposed=ServiceWorker, Constructor(ByteString method)]
interface RouterIfMethod : ServiceWorkerRouterItem {}

[Exposed=ServiceWorker, Constructor(USVString url, optional RouterIfURLOptions options)]
interface RouterIfURL : ServiceWorkerRouterItem {}

dictionary RouterIfURLOptions {
  boolean ignoreSearch = false;
}

[Exposed=ServiceWorker, Constructor(USVString url)]
interface RouterIfURLPrefix : ServiceWorkerRouterItem {}

[Exposed=ServiceWorker, Constructor(USVString url, optional RouterIfURLOptions options)]
interface RouterIfURLSuffix : ServiceWorkerRouterItem {}

[Exposed=ServiceWorker, Constructor(optional RouterIfDateOptions options)]
interface RouterIfDate : ServiceWorkerRouterItem {}

dictionary RouterIfDateOptions {
  // These should accept Date objects too, but I'm not sure how to do that in WebIDL.
  unsigned long long from = 0;
  // I think Infinity is an invalid value here, but you get the point.
  unsigned long long to = Infinity;
}

[Exposed=ServiceWorker, Constructor(optional RouterIfRequestOptions options)]
interface RouterIfRequest : ServiceWorkerRouterItem {}

dictionary RouterIfRequestOptions {
  RequestDestination destination;
  RequestMode mode;
  RequestCredentials credentials;
  RequestCache cache;
  RequestRedirect redirect;
}
```

Again, these interfaces don't have attributes, but they could reflect the options/defaults passed into the constructor.

# Shortcuts

GET requests are the most common type of request to provide specific routing for.

```webidl
partial interface ServiceWorkerRouter {
  void get(ServiceWorkerRouterItem... items);
}
```

Where the implementation is roughly:

```js
router.get = function(...items) {
  router.add(new RouterIfMethod('GET'), ...items);
};
```

We may also consider treating strings as URL matchers.

* `router.add('/foo/')` === `router.add(new RouterIfURL('/foo/'))`.
* `router.add('/foo/*')` === `router.add(new RouterIfURLPrefix('/foo/'))`.
* `router.add('*.png')` === `router.add(new RouterIfURLSuffix('.png'))`.

# Examples

## Bypassing the service worker for particular resources

```js
// Go straight to the network after 25 hrs.
router.add(
  new RouterIfDate({ to: Date.now() + 1000 * 60 * 60 * 25 }),
  new RouterSourceNetwork(),
);

// Go straight to the network for all same-origin URLs starting '/videos/'.
router.add(
  new RouterIfURLPrefix('/videos/'),
  new RouterSourceNetwork(),
);
```

## Offline-first

```js
router.get(
  // If the URL is same-origin and starts '/avatars/'.
  new RouterIfURLPrefix('/avatars/'),
  // Try to get a match for the request from the cache.
  new RouterSourceCache(),
  // Otherwise, try to fetch the request from the network.
  new RouterSourceNetwork(),
  // Otherwise, try to get a match for the request from the cache for '/avatars/fallback.png'.
  new RouterSourceCache({ request: '/avatars/fallback.png' }),
);
```

## Online-first

```js
router.get(
  // If the URL is same-origin and starts '/articles/'.
  new RouterIfURLPrefix('/articles/'),
  // Try to fetch the request from the network.
  new RouterSourceNetwork(),
  // Otherwise, try to match the request in the cache.
  new RouterSourceCache(),
  // Otherwise, if the request destination is 'document'.
  new RouterIfRequest({ destination: 'document' }),
  // Try to match '/articles/offline' in the cache.
  new RouterSourceCache({ request: '/articles/offline' }),
);
```

# Processing

This is very rough prose, but hopefully it explains the order of things.

A service worker has **routes**. The routes do not belong to the registration, so a new empty service worker will have no defined routes, even if the previous service worker defined many.

A route has **items**.

## To create a new route containing *items*

1. If the service worker has activated, throw. Routes must be created before the service worker has activated. They may be created during the install/activate event however. This might be useful if network data is needed to decide which routes to use.
1. Create a new route with *items*, and append it to **routes**.

## Handling a fetch

These steps will come before handling navigation preload, meaning no preload will be made if a route handles the request.

*request* is the request being made.

1. Let *routerCallbackId* be the empty string.
1. RouterLoop: For each *route* of this service worker's **routes**:
    1. For each *item* of *route*'s **items**:
        1. If *item* is a `RouterIfMethod`, then:
            1. If *item*'s **method** does not equal *request*'s **method**, then break.
        1. Otherwise, if *item* is a `RouterIfURL`, then:
            1. If *item*'s **url** does not equal *request*'s **url**, then break.
        1. Etc etc for other conditions.
        1. Otherwise, if *item* is a `RouterSourceNetwork`, then:
            1. Let *networkRequest* be *item*'s **request**.
            1. If *networkRequest* is null, then set *networkRequest* to *request*.
            1. Let *response* be the result of fetching *networkRequest*.
            1. If *response* is not an error, return *response*.
        1. Otherwise, if *item* is a `RouterSourceCache`, then:
            1. Let *networkRequest* be *item*'s **request**.
            1. If *networkRequest* is null, then set *networkRequest* to *request*.
            1. Let *response* be the result of looking for a match in the cache, passing in *item*'s options.
            1. If *response* is not null, return *response*.
        1. Otherwise, if *item* is a `RouterSourceFetchEvent`, then:
            1. Set *routerCallbackId* to *item*'s **id**.
            1. Break RouterLoop.
1. Call the fetch event as usual, but with *routerCallbackId* as one of the event properties.

# Extensibility

I can imagine things like:

* `RouterOr(...conditionalItems)` – True if any of the conditional items are true.
* `RouterNot(condition)` – Inverts a condition.
* `RouterIfResponse(options)` – Right now, a response is returned immediately once one is found. However, the route could continue, skipping sources, but processing conditions. This condition could check the response and break the route if it doesn't match. Along with a way to discard any selected response, you could discard responses that didn't have an ok status.
* `RouterCacheResponse(cacheName)` – If a response has been found, add it to a cache.
* `RouterCloneRequest()` – It feels like `RouterSourceNetwork` would consume requests, so if you need to do additional processing, this could clone the request.

But these could arrive much later. Some of the things in the main proposal may also be considered "v2".

**Comment by jakearchibald** [comment]
Here's an alternative model suggested by @wanderview, which is focused on allowing developers to toggle the fetch event:

```webidl
partial interface ServiceWorkerGlobalScope {
  attribute ServiceWorkerEventSubscriptions eventSubscriptions;
}

interface ServiceWorkerEventSubscriptions {
  Array<DOMString> get();
  void add(DOMString eventType);
  void remove(DOMString eventType);
}
```

Where `get` returns the [set of event types to handle](https://w3c.github.io/ServiceWorker/#dfn-set-of-event-types-to-handle).

`add` adds to the set.

`remove` removes from the set.

This means the developer would be able to expire their fetch handler after some amount of time (I believe this is Facebook's use-case).

```js
const deployTimestamp = 1543938484103;
const oneDay = 1000 * 60 * 60 * 24;

addEventListener('fetch', (event) => {
  if (Date.now() - deployTimestamp > oneDay) {
    eventSubscriptions.remove('fetch');
    return;
  }

  // …
});
```

This is a much simpler feature, but it doesn't allow skipping the service worker by route, or going straight to the cache by route.

**Comment by jakearchibald** [comment]
One more suggestion (again from @wanderview) would be to add a new CSP rule (or something similar) that defines which url-prefixes a service worker would intercept, if any.

This means the setting would be page-by-page. Navigations wouldn't be able to opt-out of service worker, but presumably navigation preload would solve a lot of performance issues there.

**Comment by jakearchibald** [comment]
cc @n8schloss, @jatindersmann, @aliams, @youennf, @asutherland

**Comment by jakearchibald** [comment]
There's a gotcha with the static routes: If you put them at top level, they'll execute without issue, but next time the service worker is started (after activating), they'll fail. Also, if the service worker starts multiple times before activation, you'll get duplicate routes.

Might make more sense to throw unless the routes are being added during the install event, so they'd always fail top-level.

Given that, it might make sense to put the API on the install event.

**Comment by annevk** [comment]
How does this relate to https://github.com/domenic/import-maps? Intuitively it feels like these should be the same thing.

**Comment by wanderview** [comment]
Just to clarify, the two suggestion I had were to support a subset of the use cases where the site might want to stop consulting the fetch handler until an update can occur.  Some sites are blocked on these use cases right now and I wondered if there was a way we could unblock them without conflicting with any future possible static routes spec.

**Comment by jakearchibald** [comment]
@annevk I think they're different enough, and can be used together. 

Import maps URLs with scheme `import:` to one or more built-in modules or HTTP fetches. Those HTTP fetches then go through the service worker.

Then, the service worker's fetch event (or its static routes) can decide how to conduct those fetches.

**Comment by jeffposnick** [comment]
> But these could arrive much later. Some of the things in the main proposal may also be considered "v2".

I can see developers eagerly adopting this approach if browsers started to implement it. Being able to string together complex routing rules with network/cache interactions without the overhead of starting up a service worker or pulling in extra runtime code (except for the polyfill scenario...) would be a nice performance win.

I have something of a meta-question about the longer term, v2+, plans though. There are some common concerns that production web apps need to think about, like cache maintenance/expiration policies for runtime caches. Today, that type of activity is likely to happen inside of a `fetch` handler. If we move to a model where you can either get

a) fast routing/response generation, but no `fetch` handler or
b) the ability to run code to deal with housekeeping, but incur the overhead of a `fetch` handler

developers might feel conflicted about that tradeoff.

One approach could be to create a new type of event that's fired after a response has been generated by the native router, like `routingcomplete`, that allows developers to run their own bookkeeping code outside of the critical response generation flow.

Looking at it from a slightly different perspective, this proposal ends up creating a native implementation of a subset of things that have up until now been accomplished using service worker runtime libraries. Cache-expiration also falls into the category of things that developers commonly accomplish via service worker runtime libraries. The feeling I got from https://github.com/w3c/ServiceWorker/issues/863 is that there hasn't been an appetite for natively implementing that as part of the Cache Storage API. If this routing proposal moves forward, does that change the equation around natively implementing higher-level functionality like cache expiration?

**Comment by jakearchibald** [comment]
That's a good point. I had this in earlier drafts but didn't think we needed it yet. Should be relatively simple to add.

**Comment by n8schloss** [comment]
This is awesome! @jakearchibald, I really like the proposal you outlined outlined in the first comment, it solves the two big issues that we're seeing 😃

  1) There's a good amount of added overhead we are measuring for fetching user generated non-cached resources when a fetch event is enabled.

  2) After a period of time the items cached in the service worker and served via the fetch event are invalid, so when starting the service worker we end up skipping the cache and fetching from the network, however we end up paying a large cost of starting the service worker in that case and get no benefit.

@jakearchibald, @wanderview's solution that you outlined in the second comment above gives us a solution to issue 2 but not issue 1. Like @wanderview said in his comment, issue 2 impacts us more right now than issue 1. So with the first proposal, as long as the spec is written in such a way that vendors can quickly implement the experary time condition without having to block on implementing the RouterIfURL conditions then I think this is really really great!

**Comment by jakearchibald** [comment]
@n8schloss thanks for the feedback! It feels like `IfURL` is pretty fundamental to other developers, but the prefix/suffix stuff could probably wait.

@jatindersmann, @aliams, @youennf, @asutherland, @mattto, @wanderview: How do you feel about this implementation-wise?

**Comment by aaronsn** [comment]
@jakearchibald - This is really great! I like the flexibility of the proposal. I’m interested in understanding better if this will extend to more complex scenarios in the future.

- Would it be possible to have something like RouterIfTimeout which activates that route but cancels if the headers aren't received within a certain timeout? This would allow for preferring the network but falling back to cache if the network takes too long.
- Can this API handle racing sources rather than doing them in sequence?
- I think RouterIfResponse is an important feature, to allow for custom handling for error responses. However, things start to get more complicated with options that apply after the request has been made. For example, say you want to prefer network if the response is a 200, then try cache, then use the server response anyway if cached response is missing. Could you accomplish that with `add(new RouterSourceNetwork(), new RouterIfResponse());
add(new RouterSourceCache(), new RouterSourceNetwork())` without it issuing multiple network requests? Or, say you want to do the same thing but if either response isn’t a 200 then use custom logic to determine which to use. Would the fetch event be able to access the responses that were already requested via the routes?
- Avoiding controlling certain sub-scopes is something I'd like to see for service workers, but I'm not sure if this API can/should provide this support. I can sort of see how you could do this with a route like `add(new RouterIfURLPrefix(), new RouterSourceNetwork())`, but you'd have to know which sub-resources will be requested. Could there be a RouterSource that causes the service worker to not control that client? Eg `add(new RouterIfURLPrefix(), new RouterSourceUnclaimClient())`. Or would that be too much of an abuse of the API?

Also one thought about implementation - if a request is handled via a static route, would the service worker still be started up in the background even though it isn't needed for the request? On the one hand, this would ensure that the service worker is started for future resource requests. On the other hand, if an entire page load (main resource plus sub resources) can be handled by static routes, then it's nice to avoid the performance cost of starting up the service worker.

**Comment by jakearchibald** [comment]
@aaronsn 

> Would it be possible to have something like RouterIfTimeout which activates that route but cancels if the headers aren't received within a certain timeout? This would allow for preferring the network but falling back to cache if the network takes too long.

I think I'd make this an option to `RouterSourceNetwork`, as it's the only one that would benefit from a timeout right now.

> Can this API handle racing sources rather than doing them in sequence?

Something like:

```js
router.get(
  new RouterSourceAny([
    ...sources
  ])
);
```

This could also have a timeout option.

> I think RouterIfResponse is an important feature

I think everything you mention here is possible, although we might be hitting edge cases that are easier to read as logic in a fetch event, and uncommon enough that optimising them via a router doesn't bring much of a benefit.

> Avoiding controlling certain sub-scopes is something I'd like to see for service workers, but I'm not sure if this API can/should provide this support

I don't think it should. Only things you can currently do in a fetch event are in scope. To do this, we either want a way to exclude routes as part of the call to `serviceWorker.register`, or a way to do it via a fetch event (after which we could look at adding something to the router).

> if a request is handled via a static route, would the service worker still be started up in the background even though it isn't needed for the request? On the one hand, this would ensure that the service worker is started for future resource requests. On the other hand, if an entire page load (main resource plus sub resources) can be handled by static routes, then it's nice to avoid the performance cost of starting up the service worker.

Interesting! The spec is deliberately loose when it comes to when the service worker is started, and how long it stays alive for. Either behaviour would be spec compatible.

If a service worker is started, and not needed, it shouldn't affect page performance as nothing's blocked on it.

**Comment by wanderview** [comment]
It might also be useful to describe the default routes you get when a service worker is installed.  Either route to FetchEvent or no where depending on if there is a fetch handler.

> @wanderview: How do you feel about this implementation-wise?

Personal opinion that does not represent any actual implementation priorities:

I guess I'm most interested in how something like this could be incrementally approached.  For example, if we started with:

1. `ServiceWorkerRouter.add()`
2. `RouterIfDate`
3. `RouterSourceNetwork` with default options

Or replace (2) with `RouterIfURL`.  This minimal initial set might unblock certain use cases.  We could then layer additional items later.  I'm not sure how people would feel about a having a partial "router" in the platform that doesn't provide a full routing capability.

To me the `RouterIfDate` is more compelling at the moment because we don't have a good alternative solution for avoiding service worker startup costs in an expired state.  `RouterIfURL` is more "router-like" but it seems oriented at carving out exceptions for certain subresources, but I feel like that is less of a problem since typically service worker startup is not necessary for subresources.

I imagine, though, there is going to be a tension between how many of these options and extensions to implement vs using javascript in FetchEvent.  For example, the list of logical combinations in the "extensibility" section seemed like perhaps something that should just be done in js.  I'm not sure we want to implement and maintain a complex DSL when we can achieve the same thing with js.

**Comment by jakearchibald** [comment]
One thing that isn't clear to me yet: 

```js
router.add(
  new RouterIfURLPrefix('/avatars/'),
  new RouterSourceCache(),
);

router.add(
  new RouterIfURLSuffix('.jpg'),
  new RouterSourceNetwork(),
);
```

If `/avatars/foo.jpg` is requested, but it isn't in the cache, what happens? Does the request fall through to the next route?

**Comment by jakearchibald** [comment]
Having slept on it, I think it's important that a single route is selected based on conditions. I'll work on a new draft that uses a `router.add(conditions, sources)` pattern.

This matches other routers like Express, where "continue to other routes" is opt-in.

**Comment by jakearchibald** [comment]
Ok, here's a second draft:

# Creating a route

**WebIDL**

```webidl
// Install currently uses a plain ExtendableEvent, so we'd need something specific
partial interface ServiceWorkerInstallEvent {
  attribute ServiceWorkerRouter router;
}

[Exposed=ServiceWorker]
interface ServiceWorkerRouter {
  void add(
    (RouterCondition or sequence<RouterCondition>) conditions,
    (RouterSource or sequence<RouterSource>) sources,
  );
}

[Exposed=ServiceWorker]
interface RouterSource {}

[Exposed=ServiceWorker]
interface RouterCondition {}
```

**JavaScript**

```js
addEventListener('install', (event) => {
  event.router.add(conditions, sources);
  event.router.add(otherConditions, otherSources);
});
```

The browser will consider routes in the order declared, and if all conditions match, each source will be tried in turn.

# Conditions

These determine if a particular static route should be used rather than dispatching a fetch event.

**WebIDL**

```webidl
[Exposed=ServiceWorker, Constructor(ByteString method)]
interface RouterIfMethod : RouterCondition {}

[Exposed=ServiceWorker, Constructor(USVString url, optional RouterIfURLOptions options)]
interface RouterIfURL : RouterCondition {}

dictionary RouterIfURLOptions {
  boolean ignoreSearch = false;
}

[Exposed=ServiceWorker, Constructor(USVString url)]
interface RouterIfURLPrefix : RouterCondition {}

[Exposed=ServiceWorker, Constructor(USVString url, optional RouterIfURLOptions options)]
interface RouterIfURLSuffix : RouterCondition {}

[Exposed=ServiceWorker, Constructor(optional RouterIfDateOptions options)]
interface RouterIfDate : RouterCondition {}

dictionary RouterIfDateOptions {
  // These should accept Date objects too, but I'm not sure how to do that in WebIDL.
  unsigned long long from = 0;
  // I think Infinity is an invalid value here, but you get the point.
  unsigned long long to = Infinity;
}

[Exposed=ServiceWorker, Constructor(optional RouterIfRequestOptions options)]
interface RouterIfRequest : RouterCondition {}

dictionary RouterIfRequestOptions {
  RequestDestination destination;
  RequestMode mode;
  RequestCredentials credentials;
  RequestCache cache;
  RequestRedirect redirect;
}
```

Again, these interfaces don't have attributes, but they could reflect the options/defaults passed into the constructor.

## Sources

These determine where the route should try to get a response from.

**WebIDL**

```webidl
[Exposed=ServiceWorker, Constructor(optional RouterSourceNetworkOptions options)]
interface RouterSourceNetwork : RouterSource {}

dictionary RouterSourceNetworkOptions {
  // A specific request can be provided, otherwise the current request is used.
  Request request;
}

[Exposed=ServiceWorker, Constructor(optional RouterSourceCacheOptions options)]
interface RouterSourceCache : RouterSource {}

RouterSourceCacheOptions : MultiCacheQueryOptions {
  // A specific request can be provided, otherwise the current request is used.
  Request request;
}

[Exposed=ServiceWorker, Constructor(optional RouterSourceFetchEventOptions options)]
interface RouterSourceFetchEvent : RouterSource {}

dictionary RouterSourceFetchEventOptions {
  DOMString id = '';
}
```

These interfaces don't currently have attributes, but they could have attributes that reflect the options/defaults passed into the constructor.

# Shortcuts

GET requests are the most common type of request to provide specific routing for.

**WebIDL**

```webidl
partial interface ServiceWorkerRouter {
  void get(/* same as add */);
}
```

Where the **JavaScript** implementation is roughly:

```js
router.get = function(conditions, sources) {
  if (conditions instanceof RouterCondition) {
    conditions = [conditions];
  }
  router.add([new RouterIfMethod('GET'), ...conditions], sources);
};
```

We may also consider treating strings as URL matchers.

* `router.add('/foo/', sources)` === `router.add(new RouterIfURL('/foo/'), sources)`.
* `router.add('/foo/*', sources)` === `router.add(new RouterIfURLPrefix('/foo/'), sources)`.
* `router.add('*.png', sources)` === `router.add(new RouterIfURLSuffix('.png'), sources)`.

# Examples

## Bypassing the service worker for particular resources

**JavaScript**

```js
// Go straight to the network after 25 hrs.
router.add(
  new RouterIfDate({ from: Date.now() + 1000 * 60 * 60 * 25 }),
  new RouterSourceNetwork(),
);

// Go straight to the network for all same-origin URLs starting '/videos/'.
router.add(
  new RouterIfURLPrefix('/videos/'),
  new RouterSourceNetwork(),
);
```

## Offline-first

**JavaScript**

```js
router.get(
  // If the URL is same-origin and starts '/avatars/'.
  new RouterIfURLPrefix('/avatars/'),
  [
    // Try to get a match for the request from the cache.
    new RouterSourceCache(),
    // Otherwise, try to fetch the request from the network.
    new RouterSourceNetwork(),
    // Otherwise, try to get a match for the request from the cache for '/avatars/fallback.png'.
    new RouterSourceCache({ request: '/avatars/fallback.png' }),
  ],
);
```

## Online-first

**JavaScript**

```js
router.get(
  // If the URL is same-origin and starts '/articles/'.
  new RouterIfURLPrefix('/articles/'),
  [
    // Try to fetch the request from the network.
    new RouterSourceNetwork(),
    // Otherwise, try to match the request in the cache.
    new RouterSourceCache(),
    // Otherwise, try to match '/articles/offline' in the cache.
    new RouterSourceCache({ request: '/articles/offline' }),
  ],
);
```

# Processing

This is very rough prose, but hopefully it explains the order of things.

A service worker has **routes**. The routes do not belong to the registration, so a new empty service worker will have no defined routes, even if the previous service worker defined many.

A route has **conditions** and **sources**.

## To create a new route containing *conditions* and *sources*

1. If the service worker is not "installing", throw. Routes must be created before the service worker has installed.
1. Create a new route with *conditions* and *sources*, and append it to **routes**.

## Handling a fetch

These steps will come before handling navigation preload, meaning no preload will be made if a route handles the request.

*request* is the request being made.

1. RouterLoop: For each *route* of this service worker's **routes**:
    1. For each *condition* of *route*'s **conditions**:
        1. If *condition* is a `RouterIfMethod`, then:
            1. If *condition*'s **method** does not equal *request*'s **method**, then continue RouterLoop.
        1. Otherwise, if *condition* is a `RouterIfURL`, then:
            1. If *condition*'s **url** does not equal *request*'s **url**, then continue RouterLoop.
        1. Etc etc for other conditions.
    1. For each *source* of *route*'s **sources**:
        1. If *source* is a `RouterSourceNetwork`, then:
            1. Let *networkRequest* be *source*'s **request**.
            1. If *networkRequest* is null, then set *networkRequest* to *request*.
            1. Let *response* be the result of fetching *networkRequest*.
            1. If *response* is not an error, return *response*.
        1. Otherwise, if *source* is a `RouterSourceCache`, then:
            1. Let *networkRequest* be *source*'s **request**.
            1. If *networkRequest* is null, then set *networkRequest* to *request*.
            1. Let *response* be the result of looking for a match in the cache, passing in *source*'s options.
            1. If *response* is not null, return *response*.
        1. Otherwise, if *source* is a `RouterSourceFetchEvent`, then:
            1. Set *routerCallbackId* to *source*'s **id**.
            1. Call the fetch event as usual, but with *source*'s **id** as one of the event properties.
            1. Return.
    1. Return a network error.
1. Call the fetch event as usual.

# Extensibility

I can imagine things like:

* `RouterOr(...conditionalItems)` – True if any of the conditional items are true.
* `RouterNot(condition)` – Inverts a condition.
* `RouterFilterResponse(options)` – Right now, a response is returned immediately once one is found. However, the route could continue, skipping sources, but processing filters. This could check the response and discard it if it doesn't match. An example would be discarding responses that don't have an ok status.
* `RouterCacheResponse(cacheName)` – If a response has been found, add it to a cache.

But these could arrive much later. Some of the things in the main proposal may also be considered "v2".

**Comment by jeffposnick** [comment]
Here's a couple of things that, based on what we've run into with Workbox's routing, tend to crop up in the real-world. I wanted to draw them to your attention so that you could think about them earlier rather than later.

### Cross-origin routing

Developers sometimes want to route cross-origin requests. And sometimes they don't. Coming up with a syntax for `conditions` that supports both scenarios can be difficult. It looks like the current proposal assumes that the URL prefix/suffix will only work for same-origin requests, so be prepared for folks asking for a cross-origin syntax at some point.

Workbox has a few different ways of specifying routing conditions, but the most common is `RegExp`-based, and we [settled on](https://developers.google.com/web/tools/workbox/guides/route-requests#matching_a_route_with_a_regular_expression) the following behavior: if the `RegExp` matches the full URL starting with the first character [the `'h'` in `'https://...'`] then we assume that it can match cross-origin requests. If the `RegExp` matches, but the match starts on anything other than the first character in the full URL, then it will only trigger if it's a same-origin request.

I would imagine folks wanting to see at least a `RouterIfURLOrigin` condition that they could use to customize this behavior, and that might need to support wildcards to deal with CDN origins that don't have fixed naming conventions.

### Non-200 OK responses sometimes need to be treated as errors

The proposal for `RouterSourceNetwork` currently reads _If response is not an error, return response._ I think some developers are going to find this too limiting. There are folks who will end up needing different behavior based on the `status` of the response, not just whether a `NetworkError` occurred.

I'm not sure what the cleanest approach is there in terms of your proposal—maybe adding in a way of setting a list of "acceptable" status codes, including letting folks opt-in or opt-out to status code `0` from an opaque response being considered a success?

**Comment by jakearchibald** [comment]
I've renamed prefix/suffix to starts/ends to match `str.startsWith`.

**Comment by jakearchibald** [comment]
@jeffposnick 

> Developers sometimes want to route cross-origin requests. And sometimes they don't.

```js
router.add(
  [
    new RouterIfURLStarts('https://photos.example.com/'),
    new RouterIfURLEnds('.jpg', { ignoreSearch: true }),
  ]
  // …
);
```

The above would match on URLs that start `https://photos.example.com/` and end `.jpg`. It gets trickier if you want to match URLs to all other origins that end `.jpg`, you'd need `RouterNot` for that.

RegExp is tricky here as there's no spec for how it could outlive JavaScript. We could try and standardise globbing, but I worry that would take a big chunk of time.

> Non-200 OK responses sometimes need to be treated as errors

I've added `rejectNotOk` as an option. Terrible name, but that's something we can bikeshed.

**Comment by nhoizey** [comment]
Hi @jakearchibald, this is really interesting!

Reading the post on your blog, I wondered how I could add the `RouterSourceNetwork` to the cache, because it looks like the `fetch` would not be fired. But I see here that you suggest adding `RouterCacheResponse` to a future "v2" version.

IMHO, this would be great in the "v1" to help build offline experiences without manually preloading.

But I understand there must be some priorities. 😅

**Comment by jakearchibald** [comment]
Yeah, we want to avoid trying to do everything at once. In terms of fetching and caching, see https://jakearchibald.com/2019/service-worker-declarative-router/#routersourcefetchevent.

**Comment by WORMSS** [comment]
This is not going to be a popular view. but I am not a massive fan of the 'string' shortcut of the conditions. I understand you want it to be as close to existing apis like express.. but not everyone understands express all the time.  
I was really enjoying the very explicit class approach and the `everything in the conditional list must eval to true`, rather than `this or thistoo or that`.

I wondered if there could be some optimisation for that, behind the scenes. Since they all have to be `true` to be considered true, then the order they are executed could be arbitrary? (I am of course assuming they are all stateless operations).

So ones that are a little more expensive to calculate could be done last, and super dirty cheap ones could be done first? Rather than `do them in the order the service worker coder wrote them in` ?? That way different browsers can optimise in their own way and what might be expensive for one is cheap for the others, and the developer doesn't need to know this when setting up the conditions.

For example,
```js
router.add(
  [
    new RouterIfURLMatchesRegex('lets assume some horrible regex'),
    new RouterIfMethod('GET'),
  ]
  // …
);
```

I know `RouterIfURLMatchesRegex` doesn't exist, but you have to believe that it would be far less expensive to do the method condition before a possible addition to the conditions in the future. Rather than try and resolve the regex, just to find it was a 'POST' request anyway, so didn't even need it.

**Comment by jakearchibald** [comment]
@WORMSS 

> I am not a massive fan of the 'string' shortcut

I see what you're saying about the string thing. Also, it isn't all that similar to express. It might be better to drop that shortcut until something like globbing can be properly spec'd.

> I wondered if there could be some optimisation for that, behind the scenes. Since they all have to be true to be considered true, then the order they are executed could be arbitrary?

In the case of multiple conditions, I'd spec them as a sequence, but if a browser decided to check them in a different order, or in parallel, there'd be no observable difference.

**Comment by yoavweiss** [comment]
My feedback:
* I've seen cases where the request destination would have been extremely helpful as a filter.
* The found the use of `get()` as a shortcut for GET requests confusing as first, as I expected it to be a way to get the set routes or something similar. Maybe it's just me, but if not, might be worthwhile to rename.

**Comment by shortercode** [comment]
Really liking the concept of avoiding the overhead of starting the ServiceWorker where possible, but I feel like not adding an additional Source type that utilises a callback is a missed opportunity. It would effectively be the same behaviour as the fetch event fallback, but specific to the given conditions. For all other source types we would still avoid starting up the service worker.

Also I think syntax wise I feel it reads better to have static methods on a RouterCondition/RouterSource class that instantiate the specific type. Although I'm not sure how in keeping that is with other web APIs.

```javascript
const { router } = e;

router.get(
  RouterCondition.startsWith("/avatars/"),
  [ RouterSource.cache(), RouterSource.network() ]
);

router.get(
  RouterCondition.fileExtension(".mp4"),
  RouterSource.network()
);

router.get(
  RouterCondition.URL("/"),
  RouterSource.cache("/shell.html")
);

router.add(
  RouterCondition.any(),
  RouterSource.custom(fetchEvent => {
    
  })
);
```

**Comment by tomayac** [comment]
@yoavweiss: I think the `.get()` is heavily inspired by Express.js’ routing: http://expressjs.com/en/guide/routing.html.

**Comment by domenic** [comment]
I apologize that this is not a very substantial contribution, but I think I'm having an allergic reaction to the "Java-esque" (or Dart-esque) nested class constructors. I'd encourage thinking about what this design would look like if done in the opposite direction: a purely JSON format. For example, something like

```js
router.route([
  {
    condition: {
      method: "GET",
      urlStartsWith: "/avatars"
    },
    destination: ["cache", "network"]
  },
  {
    condition: {
      method: "GET",
      urlPathEndsWith: ".mp4"
    }
    destination: ["network"]
  },
  {
    condition: {
      method: "GET",
      urlPath: "/"
    },
    destination: [{ cache: "/shell.html" }]
  }
]);
```

I don't think this extreme is right either (in particular the `[{ cache: "/shell.html" }]` seems weird and underdeveloped) but I think it'd be a valuable exercise to see what it would look like to have a purely declarative routing format expressed just in JS objects. Then, you could figure out where it would make sense to strategically convert some POJSOs into class instances, versus where the POJSOs are simpler.

**Comment by jakearchibald** [comment]
I'm similarly not keen on the use of classes.

As far as I know, WebIDL doesn't let an object's type be conditional on the value of a previous argument. This makes it tricky when it comes to source options. But you might be able to do it with object keys like above. I'll have a play.

The main thing I'm worried about is feature detection.

**Comment by jakearchibald** [comment]
@shortercode You probably saw my reply on my blog, but I'll post it here for others:

> I feel like not adding an additional Source type that utilises a callback is a missed opportunity

I explored that idea, but here's the issue, (spooky music) the callback ceases to exist. It only ever exists when the install event callback is called. Then later, the service worker closes, the callback is gone. Even later, when the service worker is started again, the install event callback isn't called again, so the callback still doesn't exist (and even if we called that callback again, the callback would be a different instance).

To overcome this, the callback would need to be somewhere where it'd exist every time the service worker runs, which pretty much means the top level. But how would the service worker be able to tell between two instances of the function (since a new instance is created each time the service worker is booted)? Ideally it'd be associated with something primitive, like a string. This is how events work, so you can see how I arrived at `RouterSourceFetchEvent` as the next best thing.

> I think syntax wise I feel it reads better to have static methods on a RouterCondition/RouterSource class that instantiate the specific type

I explored this too. What does `RouterCondition.startsWith("/avatars/")` return? Does that type have a constructor? That's how I landed at constructors.

**Comment by nhoizey** [comment]
> Yeah, we want to avoid trying to do everything at once.

I understand.

> In terms of fetching and caching, see https://jakearchibald.com/2019/service-worker-declarative-router/#routersourcefetchevent.

Ah, yes, sorry, I forgot about this possibility.

**Comment by richardkazuomiller** [comment]
I like this!

I often find when I'm working with changing behavior based on the URL in a Service Worker, it's easier to make an instance of `URL` and look at the parts instead of the original string. It would be nice to have conditions for the pathname, hostname, protocol, query string, etc. as well as other parts of the request. I also agree that there could be an easier way to set the condtions than a constructor.

For example:

```javascript
router.get(
  [
    // same as the RouterIfURLStarts example
    new RouterCondition.url.pathname.startsWith('/avatars/'),
    // true if url.hostname === 'example.com'
    new RouterCondition.url.hostname.is('example.com'),
    // everything should be HTTPS anyway, but since we're here
    new RouterCondition.url.protocol.is('https:'),
    // Skip this route if a queryString is set
    new RouterCondition.url.search.isEmpty(),
    // User has logged in
    new RouterCondition.request.headers.has('X-Authentication-Token'),
    // I'm sure you GET the idea
    new RouterCondition.request.method.is('GET')
  ],
  [
    // ...sources
  ],
)
```

You could also make it so conditions could be easily chained together, similar to how test frameworks do assertions.

```javascript
const profilePictureConditon = RouterCondition.url.pathname
  .startsWith('/users/')
  .endsWith('/profile.jpg');
const profilePageCondition = RouterCondition.url.pathname
  .startsWith('/users/')
  .endsWith('profile.html');
const dateCondition = RouterCondition.date
  .from(someDay)
  .to(anotherDay);
const localizedApiCondition = RouterCondition.url.hostname
  .startsWith('api.')
  .endsWith('.jp')
const globalApiCondition = RouterCondition.url.hostname
  .startsWith('api.')
  .endsWith('.com')
```

This would make code easier to read and I think makes more sense than constructors taxonomically.

All of the `RouterCondition.something.something()`s would return an instance of `RouterCondition`. To play devil's advocate in favor of constructors, I suppose it could have a constructor ...

```javascript
const conditon = new RouterCondition({
  url: {
    hostname: 'example.com',
    pathname: {
      startsWith: '/avatars/'
    },
    protocol: 'https:',
    search: ''
  },
  request: {
    method: 'GET',
    hasHeaders: ['X-Authenticaton-Token']
  }
});
```

... but if somewhere down the line a new condition gets added, I don't know how one would go about checking whether it's supported, so I'd prefer a completely constructor-less API.

**Comment by jakearchibald** [comment]
I'm not keen on the use of constructors in this proposal, but I didn't think there was another way. I'd like to explore it further, and at least show my working.

```js
router.add(conditions, sources);
```

# Conditions

This could be a simple object:

```js
router.add({
  method: 'GET',
  url: { endsWith: '.mp4', ignoreSearch: true },
  date: { to: Date.now() + 1000 * 60 * 60 * 24 * 5 },
}, sources);
```

This seems pretty nice and easy to spec. You could also add some sensible defaults like `method: 'GET'`, where the developer would need to set this to an empty string if they wanted to cover all methods.

I'm worried about feature detection. The usual trick is to give the API an object with a getter, and see which properties are read. It isn't a great user experience, and especially horrible in this case as you'd have to add a dummy route just to feature detect. I think we'd want something like:

```js
router.supportsCondition('date', { to: Date.now() }); // boolean
```

I'm not quite sure how to handle 

In terms of extensibility, here's 'or' and 'not'.

```js
router.add({
  url: { endsWith: '.mp4', ignoreSearch: true },
  or: {
    url: { endsWith: '.jpg', ignoreSearch: true },
    or: {
      url: { endsWith: '.gif', ignoreSearch: true },
    },
  },
}, sources);

// vs:

router.add(
  new RouterIfAny(
    new RouterIfURL({ endsWith: '.mp4', ignoreSearch: true }),
    new RouterIfURL({ endsWith: '.jpg', ignoreSearch: true }),
    new RouterIfURL({ endsWith: '.gif', ignoreSearch: true }),
  ),
  sources,
);
```

…and:

```js
router.add({
  not: { method: 'POST' },
}, sources);

// vs:

router.add(
  new RouterNot(new RouterIfMethod('POST')),
  sources,
);
```

I was worried it might be difficult to figure out the order of operations, but it seems to read ok, provided we can have something like `supportsCondition`.

If we support 'or' and 'not', we'd probably want 'and' too:

```js
router.add({
  method: 'POST',
  and: {
    url: {
      endsWith: '.mp4', ignoreSearch: true,
      or: {
        url: { endsWith: '.gif', ignoreSearch: true },
      },
    },
  },
}, sources);
```

# Sources

Unlike conditions, the order matters here. It could be:

```js
router.add(conditions, [
  // Where each source is a enum string:
  'network',
  // Or an array of [source, options],
  ['cache', { request: '/shell.html' }],
]);
```

It's a bit of a weird convention, but I guess using classes is weird too. But, is it possible to spec this? The sticking point seems to be `[enumValue, options]` where the type of `options` depends on the value of `enumValue`. This is pretty easy to do in TypeScript, but I don't think it's possible in WebIDL (@domenic, am I correct here?).

This has the same feature detection problem, so we'd also want:

```js
router.supportsSource('cache', { request: '/shell.html' }); // boolean
```

If the WebIDL thing becomes a sticking point, `sources` could be a sequence of enums or `RouterSource` instances:

```js
router.add(conditions, [
  'network',
  new RouterSourceCache('/shell.html'),
]);
```

If every enum had an equivalent constructor, we wouldn't need `router.supportsSource`.

# Are instances useful?

If we stuck with classes (in either a full or partial way as mentioned above), source instances could have a method to perform their action, and condition instances could have a way to test them.

```js
const condition = new RouterIfURL({ startsWith: '/article/' });
condition.test(request); // boolean
const source = new RouterSourceCache({ ignoreSearch: true });
const response = await source.doYourThing(request);
```

This might be useful when writing tests, but I'm not sure if it's useful beyond that.

# Examples

Taking the initial examples from [my blog post](https://jakearchibald.com/2019/service-worker-declarative-router/):

```js
router.get(
  new RouterIfURLStarts('/avatars/'),
  [new RouterSourceCache(), new RouterSourceNetwork()],
);

// becomes:

router.add(
  { url: { startsWith: '/avatars/' } },
  ['cache', 'network'],
);
```

```js
router.get(
  new RouterIfURLEnds('.mp4', { ignoreSearch: true }),
  new RouterSourceNetwork(),
);

// becomes:

router.add(
  { url: { endsWith: '.mp4', ignoreSearch: true } },
  'network',
);
```

```js
router.get(
  new RouterIfURL('/', { ignoreSearch: true }),
  new RouterSourceCache('/shell.html'),
);

// becomes:

router.add(
  { url: { matches: '/', ignoreSearch: true } },
  new RouterSourceCache('/shell.html'),
);

// or:

router.add(
  { url: { matches: '/', ignoreSearch: true } },
  [['cache', { request: '/shell.html' }]],
);
```

That final example looks a bit weird, so I'm leaning towards keeping classes for sources, but allowing a string in cases where options aren't needed.

**Comment by shortercode** [comment]
@jakearchibald 
> > I feel like not adding an additional Source type that utilises a callback is a missed opportunity
> 
> I explored that idea, but here's the issue, (spooky music) the callback ceases to exist. It only ever exists when the install event callback is called. Then later, the service worker closes, the callback is gone. Even later, when the service worker is started again, the install event callback isn't called again, so the callback still doesn't exist (and even if we called that callback again, the callback would be a different instance).

Ah I totally get you, mmm there's no good workaround for that really. Which is a shame. I can imagine the condition matching behaviour being duplicated in a user land library, just to provide a callback style source. I actually already have a project which could benefit from that.

> > I think syntax wise I feel it reads better to have static methods on a RouterCondition/RouterSource class that instantiate the specific type
> 
> I explored this too. What does RouterCondition.startsWith("/avatars/") return? Does that type have a constructor? That's how I landed at constructors.

I was thinking it would either return a new instance of RouterCondition, with a `{ startsWith: "/avatars/" }` configuration or an instance of RouterIfURLStarts ( which would inherit/implement RouterCondition ). RouterCondition itself would be exposed on serviceworkerglobalscope but as an illegal constructor.

If I was to polyfill the feature in JS:

```javascript
class RouterCondition {

    constructor (fn) {
        this._test = fn;
    }

    _check (fetchEvent) {
        return !!this._test(fetchEvent);
    }

    static startsWith(str) {
        return new RouterCondition(fetchEvent => {
            const url = new URL(e.request.url);
            return url.pathname.startsWith(str);
        });
    }
    
}
```

Reading feedback from @domenic  I agree that it does feel somewhat over the top to have a dozen userland classes for what is effectively immutable configuration information. Can anyone think of any methods that Conditions or Sources would actually need? The [type object pattern](http://gameprogrammingpatterns.com/type-object.html) would probably work here.

The static method suggestion I'm putting forward helps with feature detection, and stops developers needing to go **Full Java**. But I admit your latest suggestion reads very cleanly.

```javascript
router.add(
  { url: { matches: '/', ignoreSearch: true } },
  new RouterSourceCache('/shell.html'),
);
```

**Comment by jakearchibald** [comment]
@shortercode

> I was thinking it would either return a new instance of `RouterCondition`, with a { startsWith: "/avatars/" } configuration or an instance of `RouterIfURLStarts` ( which would inherit/implement `RouterCondition` ). RouterCondition itself would be exposed on serviceworkerglobalscope but as an illegal constructor.

`RouterCondition` isn't specific enough, so it'd have to be `RouterIfURLStarts`. But then, why have a method that returns an instance + an illegal constructor, when a valid constructor can do the same job. That's how I ended up using constructors.

> Can anyone think of any methods that Conditions or Sources would actually need?

Yep, I documented some above. See "Are instances useful?".

**Comment by Pajn** [comment]
This might be a bit bikeshedding but I'm not very keen on the matching of the URL as a string and would much prefer an API similar to what @richardkazuomiller proposes.
URL parsing is hard and having to disable query parameters to match on the path feels very fragile. What if there is a hash in the url for example?

**Comment by jakearchibald** [comment]
@Pajn

The difference between service worker and other routing systems is that service worker handles requests to other origins. If you create a system that only matches on URL path, would you expect that to match on any origin?

At some point it'd be nice to standardise Express-style path matching, but I don't think we'll do that as part of v1. The benefit of `{ url: { startsWith: '/articles/' } }` is we can resolve that string against the service worker's URL, as we do with URLs passed to `fetch()`.

We could also add things like `{ url: { pathStartsWith: '/foo/' } }` to match against particular components of the URL, but you'd be matching against all origins unless you specify otherwise.

> What if there is a hash in the url for example?

That's a good question. We flip-flopped on that a bit and I can't remember where we landed.

**Comment by domenic** [comment]
> This is pretty easy to do in TypeScript, but I don't think it's possible in WebIDL (@domenic, am I correct here?).

Right, you'd need to add a bit of prose. The type would be `sequence<EnumType, OptionsDictionaryType>` and then you'd use prose to validate that the sequence is length 2, and each index is as expected.

**Comment by jakearchibald** [comment]
@domenic the problem is more that the type of `OptionsDictionaryType` is dictated by the value of `EnumType`.

```js
router.add(conditions, [
  // Valid
  ['network', { request: '/' }],
  // Valid
  ['cache', { ignoreSearch: true }],
  // Invalid. ignoreSearch is not an option for the network source.
  ['network', { ignoreSearch: true }],
]);
```

**Comment by domenic** [comment]
Yeah, so either `OptionsDictionaryType` would be a union, or you would just use `object` and convert using prose.

But even then, the "invalid" should not cause an error; it just gets ignored. That's how it would work with constructor options as well.


I'll note that you could also consider the design

```js
[
  { type: 'network', request: '/' },
  { type: 'cache', ignoreSearch: true },
  { type: 'network' }
]
```

which will still require some prose I think, but is a bit less weird.

**Comment by annevk** [comment]
(Note that there's some precedent for `object` and conversion in prose. https://github.com/heycam/webidl/issues/568 has some pointers. At some point there might be enough for an abstraction to be added to IDL.)

**Comment by jeffposnick** [comment]
In case prior art is useful, [here's the syntax that Workbox uses](https://developers.google.com/web/tools/workbox/modules/workbox-build#generateSW-runtimeCaching) for expressing runtime routing information inside of a JS object. This is read by Workbox's build tools and translated into calls to the underlying Workbox libraries when generating a service worker file.

That syntax has a lot of baggage from the [earlier syntax used by `sw-precache`](https://github.com/GoogleChromeLabs/sw-precache#runtimecaching-arrayobject), which in turn had baggage due to how `sw-toolbox` was implemented. So... I'm not suggesting that it's actually something that should be used as-is. Maybe as a counter-example.

Also, I could be wrong here, but a syntax based on object properties rather than one based on classes seems like it would offer more flexibility when it comes time to write the inevitable JavaScript polyfill.

**Comment by jakearchibald** [comment]
@domenic 

> I'll note that you could also consider the design…

I generally try to avoid objects that mix required properties (`type`) with others. I realise we do this all over the platform though.

**Comment by AmeliaBR** [comment]
> I'm worried about feature detection.  [ @jakearchibald re using a configuration object instead of classes and method calls]

What if the API contained a setConfig / getConfig pair, and the getConfig drops any unrecognized properties?  So the initial server worker script could run a quick test case and then decide which configuration file to load.

Alternatively, the setConfig method could take an options parameter that defined error-checking rules: does the setter throw an error for unrecognized configuration rules/properties, or does it silently drop them, or make that rule trigger the service worker, or…?

**Comment by hewholived** [comment]
@jakearchibald I like the second proposal of using a condition object compared to list of constructors. I suggest creating RouterCondition class and a RouterPredicate class to help better express the conditions. RouterCondition class instantiates a self contained condition object that is passed to router.add function. It also exposes a test method that can be used to validate whether a request satisfies the condition. RouterPredicate class can list all the predicates that are allowed on a particular route.

```javascript
const condition = new RouterCondition((route /* of type RouterPredicate */) => {
  return route.method('GET') &&
         route.urlStartsWith('/avatar') && 
         !route.urlEndsWith('.mp4');
// return !route.method('POST');
});

router.add(condition, sources);
condition.test(request); // returns boolean; useful for other purposes
```

What do you think?

**Comment by jakearchibald** [comment]
@AmeliaBR 

> What if the API contained a setConfig / getConfig pair…

You'd also need a way to remove the route if it no longer make sense with particular features unsupported. Seems that the `supportsCondition` API might be simpler?

**Comment by jakearchibald** [comment]
@hewholived unfortunately that can't work for the same reasons as https://github.com/w3c/ServiceWorker/issues/1373#issuecomment-452222806

**Comment by richardkazuomiller** [comment]
@jakearchibald 

> We could also add things like { url: { pathStartsWith: '/foo/' } } to match against particular components of the URL, but you'd be matching against all origins unless you specify otherwise.

I think that's fine because sometimes we care about checking the domain and we don't at other times. Here's some code that's more-or-less exactly what I'm doing on a real-world service worker.


```javascript
self.addEventListener('fetch', function(event) {
  const request = event.request;
  event.respondWith(
    caches.match(request)
      .then(function(response) {
        if (response) {
          return response;
        }
        const url = new URL(request);
        if(url.pathname == '/some_path') {
          const redirectUrl = new URL(url.href);
          redirectUrl.pathname = '/';
          return Response.redirect(redirectUrl.href,302);
        }
        if(url.hostname == IMAGE_CDN_HOSTNAME) {
          return fetchFromImageCdn(request);
        }
        return fetch(request);
      })
  );
});

function fetchFromImageCdn (request) {
  // basically fetch once and cache forever
}
```

Technically, both `https://mywebsite.foo/some_path` and `https://myimagecdn.foo/some_path` will both redirect to `/`, but I know that the page will never try to access `https://myimagecdn.foo/some_path` so it's fine that both match.

I'd like to use declarative routing to do something like this

```javascript
// Things that try the cache, then go to the network
const cacheThenNetwork = [RouterSource.cache(), RouterSource.network()];

// These paths match all domains, but I'm only using one so it's fine
router.get(RouterCondition.url.pathname.is('/'), cacheThenNetwork)
router.get(RouterCondition.url.pathname.startsWith('/static/'), cacheThenNetwork)
// Any path on the image CDN
router.get(RouterCondition.url.hostname.is(MY_IMAGE_CDN), cacheThenNetwork)

// fetch event listener can now be simplified
self.addEventListener('fetch', function(event) {
  const request = event.request;
  const url = new URL(request);
  if(url.pathname == '/some_path') {
    const redirectUrl = new URL(url.href);
    redirectUrl.pathname = '/';
    event.respondWith(Response.redirect(redirectUrl.href,302));
  }
});
```

**Comment by jakearchibald** [comment]
> but I know that the page will never try to access `https://myimagecdn.foo/some_path` so it's fine that both match.

That seems really fragile vs:

```js
if (url.origin === location.origin && url.pathname == '/some_path') {
  // …
}
```

I wouldn't want developers to be matching a path on all origins unless it's explicit.

**Comment by richardkazuomiller** [comment]
You're right; that would be better. My point is not so much that it's a *good* thing to have a path matching multiple domains, but having something like `pathStartsWith` isn't necessarily creating a new problem.

I know I could get my example to work using `RouterIfURLStarts` and `RouterIfURL` (or whatever equivalent API you decide to go with), so I think the conditions in your current draft are sufficient for a v1, but having conditions for each part of a request (`href`, `protocol`, `hostname`, `pathname`, `search`, `method`, and `headers`) would be nice to have.

To go even deeper, I'd love to be able to do something like 

```javascript
router.add(
  new RouterNot(new RouterConditionHasCookie('I_AM_LOGGED_IN')),
  new RouterSourceCache({ request: '/login.html' })
)
```

**Comment by jakearchibald** [comment]
After thinking about it, I think I'm going to go with the object based approach & @domenic's suggestion:

```js
// Without options:
router.add(
  conditions,
  'cache',
);

// With options:
router.add(
  conditions,
  { type: 'cache', request: '/shell.html' },
);

// Multiple sources:
router.add(
  conditions,
  [
    'cache',
    'network',
    { type: 'cache', request: '/shell.html' },
  ],
);
```

I'm still not keen on having objects with a required property that determines the type of the object, but it seems better than the alternatives.

**Comment by WORMSS** [comment]
With the object based approach, how to you specify that you want 'cache' or 'network' as race?

**Comment by jakearchibald** [comment]
@WORMSS 

```js
router.add(
  conditions,
  { type: 'race', sources: […sources]},
);
```

The pattern feels equally extensible.

**Comment by bkardell** [comment]
fwiw, @domenic and others expressed much better what I briefly tried to say yesterday in some other venue (twitter maybe?) - I like this recent turn though quite a bit.

**Comment by Siilwyn** [comment]
The option explored in https://github.com/w3c/ServiceWorker/issues/1373#issuecomment-452247362 feels more straightforward and in this case better than using classes, would definitely prefer this version!

**Comment by jeremy-coleman** [comment]
I'd like to suggest a somewhat different approach that I feel addresses the same issue in a more intuitive manner, addresses more use cases, and easier to adopt for an average developer, which is to provide a (virtual) file system to code against the assets as they will be on the client and in the shape/structure they will actually be in. 

![image](https://user-images.githubusercontent.com/25135855/51168082-44251800-1876-11e9-9924-ed7180c9d513.png)

**Comment by jakearchibald** [comment]
@jeremy-coleman The web isn't compatible with a filesystem. It's compatible with a request/response store, which is what the cache API provides. However, this alone can't communicate when a request should go to the network. I guess I don't understand your proposal fully.

**Comment by jeremy-coleman** [comment]
I guess from my perspective, the cached files are a file system. Instead of (req res) => fetchHttpRoute => interceptEverything someLogic => if(cache) else(usenet) => next

Req res => fetchCacheRoute => if(!nocache) fetchHttp => next

If you explicitly write the api to query the cached files first you can just drop the intercept all together

**Comment by jakearchibald** [comment]
@jeremy-coleman in your system, how do you express "For any URL path ending `.jpg`, try to fetch from the network, otherwise fall back to this generic image…"?

**Comment by jeremy-coleman** [comment]
it'd be the same for both online and offline. Assuming all routes point to the cache firs - if it's an online asset, replace with net when available. same for offline, just the timeout for the cache with online data would be 0ms / fetch onSomeUserEvent compared something like a 24 hour reset for offline stuff. I think a harder thing to overcome would be designating stuff as something like 'use last successful' vs 'use constant fallback' , Proxy.revoke() with your conditional checks on propkeys access could probably handle everything needed. 

```js
router.get(
  new RouterIfURLPrefix('/**/*.jpg/'),   // If the URL ends in jpg'.
  new RouterSourceCache(),   //Otherwise, try to match the request in the cache.
  new RouterSourceNetwork(),   //Try to fetch the request from the network.   <-- this doesn't need to come first for online reqs , just use the current value in the cache first and always
```
somewhat unrelated but really what underpins my line of thinking is that I feel like offline apps should basically be completely downloaded into some form of local storage on 1st req, and the SW routes should support coding against the offline assets more-so than online sources.

**Comment by jakearchibald** [comment]
@jeremy-coleman based on your example above, I don't understand the difference between your proposal and mine.

**Comment by asakusuma** [comment]
A bit late to the party but...

+1 for the "side effect operation" issue raised by @jeffposnick and @nhoizey

In addition to refreshing the cache based on a network source, firing analytics beacons is another  feature that would be heavily used by LinkedIn. We have a lot of instrumentation to measure how the service worker is working/not working. The highest priority V2 feature for us would be support for a `routingcomplete` or similar event, to handle both of these cases.

@jakearchibald 
> I think I'd make this an option to RouterSourceNetwork, as it's the only one that would benefit from a timeout right now.

I think the timeout feature would be useful for `RouterSourceFetch`, as it might address https://github.com/w3c/ServiceWorker/issues/1292.

Timeout support would be really nice for creating a global fallback "catch all" handler.
```
router.get(
	new RouterIfURLStarts(‘/profile/*‘),
	[new RouterSourceCache(), new RouterSourceNetwork({ timeout: 10000 })]
);

router.get(‘*’, new RouterSourceCache(‘/oops.html’))
```

**Comment by jakearchibald** [comment]
@n8schloss something I wanted to double check about this proposal: A service worker's routes live with the service worker. They can't be changed without shipping a new service worker. Does that work for you?

I ask because with navigation preload, you wanted to change things during the life of a service worker (the header value).

**Comment by n8schloss** [comment]
> @n8schloss something I wanted to double check about this proposal: A service worker's routes live with the service worker. They can't be changed without shipping a new service worker. Does that work for you?
> 
> I ask because with navigation preload, you wanted to change things during the life of a service worker (the header value).

Yep! As long as there's the RouterIfDate options then our use case here will be met :)

**Comment by mgiuca** [comment]
Late to the party but I'd like to give a bit of feedback.

I love the general concept. Seems like this will solve a lot of problems (speed issues with spinning up SWs, and the added "scariness" of what if the fetch handler has a bug that makes my site non-updateable.) This came up as a potential solution to w3c/manifest#774.

I have some superficial criticism (API surface details).

* I'd like to remove the `startsWith` and `endsWith` things in favour of just having a glob syntax. Fewer API calls, less verbose syntax, and more flexible.
* `ignoreSearch` is a confusing name (why not `ignoreQuery`)? Alternatively, just embrace the glob and allow `?*` at the end of the glob to mean "ignore the query" (not as an especially special case, but rather as a general rule, treat a '?' at the end of a path with nothing after it as the same as no '?'.
* The name `router.get` is confusing because a method called "get" implies it's going to return some information out of the router, not set up a new route. I'd rather just remove that method and just make you write `router.add({method: 'GET'})` which is nice and obvious what it's going to do.

**Comment by WORMSS** [comment]
Ignore search is because it's what the JavaScript code calls it In `window.location`

**Comment by jakearchibald** [comment]
> * I'd like to remove the `startsWith` and `endsWith` things in favour of just having a glob syntax. Fewer API calls, less verbose syntax, and more flexible.

This has come up a few times in this issue. I'm not against adding it at some point, but it feels like a big contentious thing to standardise for v1.

> * `ignoreSearch` is a confusing name (why not `ignoreQuery`)?

As @WORMSS says, it's for consistency with the rest of the platform. `url.search`, `cache.match(url, { ignoreSearch: true })` etc etc.

> * The name `router.get` is confusing because a method called "get" implies it's going to return some information out of the router,

Yeah, maybe. Although this is how most node routers seem to do it.

**Comment by mgiuca** [comment]
@WORMSS Yeah it's called that in the URL object too. It's called _query_ inside the spec language but now that I re-read it, I realised that the word "query" never appears in the API interface itself. So "search" is fine.

**Comment by fallaciousreasoning** [comment]
Curious whether we could use @wanderview's URLPattern proposal instead of `startsWith`/`endsWith`.

**Comment by wanderview** [comment]
I have run into a case where declarative routing avoiding "FetchEvent dispatch for resources that will always hit network" could help:

https://bugs.chromium.org/p/chromium/issues/detail?id=1293176#c11

In this case a dev environment with unbundled modules is triggering 1000+ requests on load.  The FetchEvent processing is maxing out the SW thread.

Just offering this as a data point about where performance benefits might actually start to kick in for this kind of feature.

**Comment by yoshisatoyanagisawa** [comment]
FYI, we are about to implement a subset of the API.
https://github.com/yoshisatoyanagisawa/service-worker-static-routing-api

**Comment by webmaxru** [comment]
Thanks for your work!
FYI: I tweeted about it to get the developer community ready for experimenting :)
https://twitter.com/webmaxru/status/1664214464999182338

**Comment by sisidovski** [comment]
# Should we allow registerRouter() to be called multiple times?

As commented in https://github.com/w3c/ServiceWorker/issues/1373#issuecomment-1569530880, we are about to implement a subset, as ServiceWorker Static Routing API.
https://github.com/yoshisatoyanagisawa/service-worker-static-routing-api

We (Google Chrome team) are ready for [the Origin Trial](https://developer.chrome.com/origintrials/#/view_trial/1458040379361198081), it's available from M116. In the meantime, we'd like to hear opinions around how `InstallEvent.registerRouter()` should work in this API.

Unlike add() or get() in the original Jake's proposal, `registerRouter()` sets all the routes at once, and it can only be called once. This is for ease of understanding the latest routes. However, we may rethink this limitation because we saw [some interests](https://twitter.com/EWErikWitt/status/1684239502322151431) in using this API from 3rd parties. 3rd parties in this context mean the SW scripts which are served from cross origins, and imported by the main SW script via importScripts() or ES modules.

Like [Speed Kit](https://www.speedkit.com/), some companies provide ServiceWorker scripts as SDK, and customer websites use their scripts via importScripts(). If both a 3rd party provider and a customer want to use registerRouter(), in our current implementation the browser only registers the routing info called for the first time, and throws type error for the second registerRouter() call.

I personally feel it makes sense that 3rd parties use registerRouter() to handle something, but do you think we should support multiple registerRouter() calls? If so, a naive approach is just adding a new routing rule into existing rules, but do we need additional mechanisms or API surfaces to manage registered routes more smartly?

cc: @ErikWitt

**Comment by domenic** [comment]
I think allowing multiple calls is fine. In that case I would rename the method from `registerRouter()` to `registerRoutes()`.

The naive approach makes perfect sense to me. In other words, `registerRoutes(a); registerRoutes(b);` should be equivalent to `registerRoutes([...a, ...b]);`. I think that is what everyone would expect.

**Comment by jakearchibald** [comment]
Given that, is there a benefit to `registerRoutes` vs calling `registerRoute` as many times as needed? One route at a time would make it easier to tell which route caused a throw.

**Comment by ErikWitt** [comment]
Heythere, sorry for being so late to the discussion. My two cents:

1. Calling registerRoute from multiple handlers would be awesome to make combining multiple Service Workers easier. We have that case from time to time and it can be a complex task.

2. Calling registerRoutes outside the install event is off the table right? It's not a deal breaker but would have been convenient. At the moment out Service Worker loads a configuration from indexeddb (the service worker is generic, every customer has a unique config), so it would be convenient to configure the router outside the install event. That said, we are looking into inlining the config in our shipped service worker for every customer which would resolve that issue for us.

3. Would it be possible to make the origin trial a 3rd party origin trial? I.e. could send the origin trial token in the 3rd party script imported into the same origin service worker? If it is not a third part origin trial, our customers would need to implement the header themselves which is a lengthy process for large e-commerce companies.

btw. I love the use of URLPattern in the api :) Looking forward to trying this in production and reporting back on the performance gains

**Comment by sisidovski** [comment]
Thank you all for the feedback!

@jakearchibald That's a fair point. WDYT @yoshisatoyanagisawa ?

@ErikWitt 
> Calling registerRoutes outside the install event is off the table right?

At the moment we want to keep it inside the install event. We don't want to make the API dynamically configurable. If you have a workaround please consider doing so.

> Would it be possible to make the origin trial a 3rd party origin trial?

OriginTrial for ServiceWorker features is running on a bit different mechanism, and unfortunately 3rd party origin trial is not supported yet. We understood the use case from 3rd party context. Let us consider how to deal with it, but please expect it takes some time to solve.

**Comment by sisidovski** [comment]
I came up with this scenario, perhaps it can be a problem for some cases.

**a.com/sw.js:**
```js
importScripts('b.com/imported-sw.js');

addEventListener('install', (event) => {
  event.registerRouter({
    condition: {
     // example.jpg is returned form the fetch handler.
      urlPattern: {pathname: "example.jpg"}
    },
    source: "fetch-event"
  });
})
```

**b.com/imported-sw.js:**
```js
addEventListener('install', (event) => {
  event.registerRouter({
    condition: {
     // All jpg resources are fetched from the network.
      urlPattern: {pathname: "*.jpg"}
    },
    source: "network"
  });
})
```

`registerRouter()` in imported-sw.js is executed first, and then the one in sw.js is executed. In the current algorithm, the API simply has the list of registered routing info and try to evaluate from the first item. So any requests to jpg resources, including `example.jpg` are matched with `{pathname: "*.jpg"}` which is registered in the imported-sw.js and the routing info registered by the main sw.js is never used.

Do you think the API should have a mechanism to address this case which is introduced by allowing multiple `registerRoutes` or `registerRoute` calls? I think this is kind of a WAI behavior, but love to hear anyone's thoughts around it.

**Comment by domenic** [comment]
I think this behavior is WAI. If you call `importScripts()` on a script before your own `install` handler, you are giving that script priority to install routes.

If you would like to have your own code take priority, then you should rearrange `sw.js` like so:

```js
addEventListener('install', (event) => {
  event.registerRouter(/* ... */);
});

importScripts('b.com/imported-sw.js');
```

**Comment by jakearchibald** [comment]
It could be `addRoutes(...routes)`.

I usually don't like that pattern, since it prevents adding new parameters in future, but if seems like any options would be per route.

`add` is shorter than `register`, and `add` makes it clearer that it's additive to the previous call. It might not make it clear that previous routes are cleared with the installation of a new service worker, but I'm not sure `register` makes that clear either.

**Comment by yoshisatoyanagisawa** [comment]
FYI, we are discussing on the way to handle an empty "not" condition in https://github.com/WICG/service-worker-static-routing-api/issues/22.  We are leaning on raising.  Please feel free to share your opinions.

**Comment by quasi-mod** [comment]
FYI, we are currently actively discussing extending the resource timing API to support static routing API.
https://github.com/w3c/resource-timing/issues/389

**Comment by quasi-mod** [comment]
FYI, we’ve started the Chrome origin trial from M131 to add the proposed fields to resource timing api.
If you are interested, please participate from the origin trial dashboard [link](https://developer.chrome.com/origintrials/#/view_trial/1689412810217357313).
We would really appreciate your insightful feedback!

**Comment by webmaxru** [comment]
> FYI, we’ve started the Chrome origin trial from M131 to add the proposed fields to resource timing api. If you are interested, please participate from the origin trial dashboard [link](https://developer.chrome.com/origintrials/#/view_trial/1689412810217357313). We would really appreciate your insightful feedback!

Thank you for this update! I registered origin trial both for Chrome and Edge for one of my side projects with 1K+ monthly audience

**Comment by quasi-mod** [comment]
> > FYI, we’ve started the Chrome origin trial from M131 to add the proposed fields to resource timing api. If you are interested, please participate from the origin trial dashboard [link](https://developer.chrome.com/origintrials/#/view_trial/1689412810217357313). We would really appreciate your insightful feedback!
> 
> Thank you for this update! I registered origin trial both for Chrome and Edge for one of my side projects with 1K+ monthly audience

@webmaxru Thanks so much for registering! Looking forward for your feedbacks :)
